src/lightning/fabric/strategies/model_parallel.py

```diff
@@ -0,0 +1,331 @@
+# Copyright The Lightning AI team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import itertools
+from contextlib import ExitStack
+from datetime import timedelta
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Callable, ContextManager, Dict, Literal, Optional, TypeVar, Union
+
+import torch
+from lightning_utilities.core.rank_zero import rank_zero_only as utils_rank_zero_only
+from torch import Tensor
+from torch.nn import Module
+from torch.optim import Optimizer
+from typing_extensions import override
+
+from lightning.fabric.plugins import CheckpointIO
+from lightning.fabric.plugins.collectives.torch_collective import default_pg_timeout
+from lightning.fabric.strategies.fsdp import (
+    _distributed_checkpoint_load,
+    _distributed_checkpoint_save,
+    _has_meta_device_parameters_or_buffers,
+)
+from lightning.fabric.strategies.launchers.subprocess_script import _SubprocessScriptLauncher
+from lightning.fabric.strategies.parallel import ParallelStrategy
+from lightning.fabric.strategies.strategy import TBroadcast, _BackwardSyncControl
+from lightning.fabric.utilities.distributed import (
+    ReduceOp,
+    _distributed_is_initialized,
+    _get_default_process_group_backend_for_device,
+    _init_dist_connection,
+    _sync_ddp_if_available,
+)
+from lightning.fabric.utilities.distributed import group as _group
+from lightning.fabric.utilities.imports import _TORCH_GREATER_EQUAL_2_3
+from lightning.fabric.utilities.rank_zero import rank_zero_only, rank_zero_warn
+from lightning.fabric.utilities.seed import reset_seed
+from lightning.fabric.utilities.types import _PATH
+
+if TYPE_CHECKING:
+    from torch.distributed.device_mesh import DeviceMesh
+
+TModel = TypeVar("TModel", bound=Module)
+
+
+class ModelParallelStrategy(ParallelStrategy):
```

Comment:
I find this name confusing. Data-parallelism is not model parallelism yet this class supports both. Can we think of something else?

What about `ManualParallelStrategy`? It was the name proposed a million years ago: https://github.com/Lightning-AI/pytorch-lightning/issues/11922

Reply:
Some discussion internally to consider another name, will rename the strategy in a follow up if we converge to a final decision.